accessor/accessor_helper.hpp

```diff
@@ -59,49 +59,47 @@ namespace detail {
 
 /**
  * This helper runs from first to last dimension in order to compute the index.
- * The index is computed like this:
- * indices: x1, x2, x3, ...
+ * The index is expected to be of type `IndexType`, and will be computed in
+ * `IndexType`.
+ * The index is computed like this: indices: x1, x2, x3, ...
  * compute(stride, x1, x2, x3) -> x1 * stride[0] + x2 * stride[1] + x3
  */
-template <typename ValueType, size_type total_dim, typename DimensionType,
-          size_type current_iter = 1>
+template <typename IndexType, size_type total_dim, size_type current_iter = 1>
 struct row_major_helper_s {
     static_assert(total_dim >= 1, "Dimensionality must be >= 1");
     static_assert(current_iter < total_dim, "Iteration must be < total_dim!");
 
     static constexpr size_type dim_idx{current_iter - 1};
 
-    template <typename FirstType, typename... Indices>
-    static constexpr GKO_ACC_ATTRIBUTES ValueType
-    compute(const std::array<DimensionType, total_dim>& size,
-            const std::array<ValueType, (total_dim > 1 ? total_dim - 1 : 0)>&
-                stride,
-            FirstType first, Indices&&... idxs)
+    template <typename SizeType, typename... Indices>
+    static constexpr GKO_ACC_ATTRIBUTES IndexType compute(
+        const std::array<SizeType, total_dim>& size,
```

Comment:
Actually, `IndexType` and `SizeType` are currently different types (in a different version, they might be unified). `IndexType` is the type for the indices, and `SizeType` is the type of the dimension and stride.
Inside Accessors, you (somewhat need) an `std::array` containing the max sizes (basically the non-Ginkgo alternative to `gko::dim<>`) and for most of them, another `std::array` containing the strides.
Currently, both the dim and the strides use `size_type` internally, which is different of `IndexType`.
The reason why I made `IndexType` explicit (part of the struct and not a template parameter) was to:
1. Force the input to actually be an integer type and not an arbitrary type
2. Unify the index computation. Before, it was possible to have 32bit and 64bit inputs mixed, which performed quite badly on some architectures.

However, it might make sense to immediately move to `int64_t` being the `size_type` inside the accessor since it is not a public interface. I will look what we need to do in order to achieve that.